### PR TITLE
run script in local docker container

### PR DESCRIPTION
**Pull request by b01901143**
Added README file.

Two running options:
- running directly with commandline.
- running as docker image.

### CODE REVIEW

File: Dockerfile
Comment (spiffxp) [inline Dockerfile]: Prefer `$()` to backticks
```suggestion
RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && \
```

File: Dockerfile
Comment (spiffxp) [inline Dockerfile]: Consider installing from a versioned archive instead of using a distro's package.  This makes it easier to get components.  e.g. https://github.com/kubernetes/test-infra/blob/master/images/gcloud/Dockerfile#L36-L45

Won't block on this

File: Dockerfile
Comment (spiffxp) [inline Dockerfile:28]: nit: this is going to copy `entrypoint.sh` to `/usr/src/app` when you're also copying it to `/usr/local/bin` below

won't block on this

File: Dockerfile
Comment (spiffxp) [inline Dockerfile]: Prefer the ENV statement for setting env vars (ref: https://docs.docker.com/engine/reference/builder/#env)
```suggestion
ENV GOOGLE_APPLICATION_CREDENTIALS "/usr/src/app/gcloud_key.json"
```

File: Dockerfile
Comment (spiffxp) [inline Dockerfile]: I'm not sure what the intent is here.  I would suggest using ENTRYPOINT or CMD but not both.

File: Dockerfile
Comment (spiffxp) [inline Dockerfile]: backwards compat with what? I would probably just 
```suggestion
COPY entrypoint.sh /
```

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: Technically you don't need this because of the ENV in the Dockerfile.  I would follow convention of other env's I've suggested here, and default this if not already set.

File: README.md
Comment (spiffxp) [inline README.md]: I'll need to mount in the service account key somehow I think?

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: I would suggest parameterizing this as an env var with a default, e.g.
```suggestion
GCP_PROJECT="${GCP_PROJECT:-k8s-jkns-gke-soak}"
gcloud config set project "${GCP_PROJECT}"
```

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: Same parameterizing on zone here (note: I would put the env var declarations at the top of the file
```suggestion
GCP_ZONE="${GCP_ZONE:-us-central1-f}"
gcloud config set compute/zone ${GCP_ZONE}
```

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: In general I would avoid doing this inside a container.  In this case, you don't need to do this because you're activating the service account below

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: Prefer quoting bracing variables (ref: https://google.github.io/styleguide/shellguide.html#variable-expansion)
```suggestion
gcloud auth activate-service-account --key-file "${GOOGLE_APPLICATION_CREDENTIALS}"
```

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: I would parameterize this as well

File: entrypoint.sh
Comment (spiffxp) [inline entrypoint.sh]: nit: I would make this script executable and invoke it directly
```suggestion
./secret-script.py --create --secret_id=docker-secret --file=test-secret.yaml
```

File: secret-script.py
Comment (spiffxp) [inline secret-script.py:1]: Allow this to be directly executable via whichever `python3` is available first in PATH
```suggestion
#!/usr/bin/env python3
```

File: secret-script.py
Comment (spiffxp) [inline secret-script.py]: General CLI note: this interface is a little unintuitive.  

It's presented as a bunch of flags but only one of them will take effect, and the order of precedence is somewhat arbitrary.  Consider instead an interface where there is a "verb" that you could use in a case statement, both for execution and further validation.

Also, FWIW, the convention I have seen in kubernetes/test-infra python is to use `argparse` to avoid taking on an external dependency (e.g. https://github.com/kubernetes/test-infra/blob/master/triage/summarize.py#L587-L596)

File: secret-script.py
Comment (spiffxp) [inline secret-script.py]: Personally I would keep this script non-interactive, but if you find the input useful I would at least ask for a `--force` flag to be able to override it

File: Dockerfile
Comment (spiffxp) [reply Dockerfile]: After reading through this PR I would suggest
```suggestion
CMD ["/entrypoint.sh"]
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some comments and suggestions.  I didn't get as far as trying to build and run this locally since it's hardcoded to point at your project/cluster/acct etc.

File: README.md
Comment (spiffxp) [inline README.md]: For followup (not this PR) I would recommend naming this after a gcr.io repo you have push access to, since that will be the next step for running on kubernetes, e.g. maybe `gcr.io/k8s-jkns-gke-soak/secret-script` to match the name of the project you're using and the repo the image is built from

File: Dockerfile
Comment (b01901143) [reply Dockerfile]: Fixed as suggested.

File: Dockerfile
Comment (b01901143) [reply Dockerfile]: Agreed. Fixed as suggested.

File: Dockerfile
Comment (b01901143) [reply Dockerfile:28]: Avoided copying entrypoint.sh twice.

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: I forgot to delete this line.
done.

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: done

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: done

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: That is true.
Removed.

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: done

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: done

File: entrypoint.sh
Comment (b01901143) [reply entrypoint.sh]: done

File: secret-script.py
Comment (b01901143) [reply secret-script.py:1]: Amazed.
Fixed.

File: secret-script.py
Comment (b01901143) [reply secret-script.py]: revised to using an "action" argument including options: "create", "get", "update", and "delete".

File: secret-script.py
Comment (b01901143) [reply secret-script.py]: revised to be non-interactive, always sync'ing

File: README.md
Comment (b01901143) [reply README.md]: Currently I have the key located locally in the working directory along with other src files.
Thus it would copied to WORKDIR in the container through this line: https://github.com/b01901143/secret-script/blob/fc570896c66c02c7218d0bde14f696dbdfc6c23d/Dockerfile#L28 
Is there a preferred way other than this?

Comment (spiffxp) [reply README.md]: There is; don't store it in the container, mount it in as a volume.  It is generally not a good idea to store secrets inside of a container, because it makes it easy to accidentally leak them, and rotating them requires a rebuild of the container.

```sh
# put service account key in a directory since docker doesn't mount files
mkdir -p service-account && mv spiffxp-sa-key.json service-account

# -v src:dst:mode; in this case
#   - use $(pwd) to generate absolute path for docker
#   - dst doesn't have to already exist
#   - use ro to make volume readonly
# -e ENV=value; in this case demonstrating how I'm overriding the default set in the Dockerfile
docker run -ti --rm \
  -v $(pwd)/service-account:/tmp/service-account:ro \
  -e GOOGLE_APPLICATION_CREDENTIALS=/tmp/service-account/spiffxp-sa-key.json \
  gcr.io/spiffxp-gke-dev/secret-script \
  /bin/bash -c 'gcloud auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS'
# returns 'Activated service account credentials for: [spiffxp-demo-sa@spiffxp-gke-dev.iam.gserviceaccount.com]'
```

File: secret-script.py
Comment (spiffxp) [reply secret-script.py:1]: Ah right, `./script.py` won't work until you `chmod +x` this file

File: k8s_utils.py
Comment (spiffxp) [inline k8s_utils.py]: since you're using argparse now, you don't need these absl imports

File: gcloud_utils.py
Comment (spiffxp) [inline gcloud_utils.py]: don't need absl imports here as well

File: requirements.txt
Comment (spiffxp) [inline requirements.txt]: and you can remove absl from requirements

File: README.md
Comment (spiffxp) [inline README.md]: What permissions does the service account need? When I ran this with a freshly created service account I ran into a bunch of errors.

File: secret-script.py
Comment (spiffxp) [inline secret-script.py]: It would be useful for demo purposes to describe what you're syncing from and to

File: secret-script.py
Comment (spiffxp) [inline secret-script.py]: FWIW I more commonly see `parser.parse_args()` which eliminates some complexity in the function signature (you could just call `main(parse_args())` instead of passing in a slice, for example)

**Review by spiffxp — CHANGES_REQUESTED** [review]
This is pretty close

File: README.md
Comment (b01901143) [reply README.md]: Added to README.

Will need container.secrets.* and container.clusterViewer permissions.

File: k8s_utils.py
Comment (b01901143) [reply k8s_utils.py]: done

File: gcloud_utils.py
Comment (b01901143) [reply gcloud_utils.py]: removed

File: requirements.txt
Comment (b01901143) [reply requirements.txt]: removed

File: secret-script.py
Comment (b01901143) [reply secret-script.py]: Didn't notice this. You're right.
Fixed.

File: secret-script.py
Comment (b01901143) [reply secret-script.py]: Add from and to in the synchronizing message.

File: README.md
Comment (spiffxp) [inline README.md]:
```suggestion
			- service-secret-role.yaml
```

File: README.md
Comment (spiffxp) [inline README.md]:
```suggestion
				    gcloud iam roles create <iam-role-id> --project=k8s-jkns-gke-soak --file=service-secret-role.yaml
```

File: README.md
Comment (spiffxp) [inline README.md]:
```suggestion
				    title: Kubernetes Engine Secret Admin
				    description: Provides access to management of Kubernetes Secrets
```

File: README.md
Comment (spiffxp) [reply README.md]: consider including this file in the repo

File: README.md
Comment (spiffxp) [inline README.md]:
```suggestion
				    gcloud iam roles create <iam-role-id> --project=k8s-jkns-gke-soak --file=service-secret-role.yaml
```

File: README.md
Comment (spiffxp) [inline README.md:9]: You'll need `roles/secretmanager.admin` to work with Secret Manager secrets

**Review by spiffxp — CHANGES_REQUESTED** [review]
One more tiny round

File: README.md
Comment (b01901143) [reply README.md]: Added. Though in this demo I chose the other method (roles/container.developer), because I do not have the permission to add a custom role.

File: README.md
Comment (b01901143) [reply README.md:9]: missed that, added.

File: README.md
Comment (b01901143) [reply README.md]: revised

File: README.md
Comment (b01901143) [reply README.md]: Much better

**Review by spiffxp — APPROVED** [review]
LGTM

**Pull request merged**
